### PR TITLE
Major update with Sub&Crab Keyboard mechanics implemented.

### PR DESCRIPTION
**Pull request by CCD729**
Await for reviews .

### CODE REVIEW

File: Assets/TestScript/BulletMovement.cs [hunk 1]
```diff
@@ -7,10 +7,16 @@ public class BulletMovement : MonoBehaviour
...
+
+        foreach (GameObject sub in subs)
+        {
+            Physics.IgnoreCollision(sub.GetComponent<Collider>(), GetComponent<Collider>());
+        }
+        foreach (GameObject crab in crabs)
         {
-            Physics.IgnoreCollision(redship.GetComponent<Collider>(), GetComponent<Collider>());
+            Physics.IgnoreCollision(crab.GetComponent<Collider>(), GetComponent<Collider>());
         }
```
Comment (SeijiEmery) [inline Assets/TestScript/BulletMovement.cs]: remove these lines. if collision occurs, move the projectile in front of the barrel, or remove the collider on the barrel, etc. Or add physics layers in your test scene: https://docs.unity3d.com/Manual/LayerBasedCollision.html. we'll have a better solution for this long term, but we don't want these lines in this script.

File: Assets/TestScript/CrabTestController.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            transform.Translate(Vector3.right * speed * Time.deltaTime, Space.World);
+        if (Input.GetKey(KeyCode.UpArrow))
+            transform.Translate(Vector3.forward * speed * Time.deltaTime, Space.World);
+        if (Input.GetKey(KeyCode.DownArrow))
+            transform.Translate(Vector3.back * speed * Time.deltaTime, Space.World);
+        if (Input.GetKey(KeyCode.Q))
+            transform.Rotate(Vector3.down * rotateSpeed, Space.Self);
+        if (Input.GetKey(KeyCode.E))
+            transform.Rotate(Vector3.up * rotateSpeed, Space.Self);
+        if (Input.GetKey(KeyCode.Tab))
```
Comment (SeijiEmery) [inline Assets/TestScript/CrabTestController.cs]: would ideally be using GetAxis + GetButton but this is fine

File: Assets/TestScript/BulletMovement.cs (hunk 1 above)
Comment (SeijiEmery) [reply Assets/TestScript/BulletMovement.cs]: BulletMovement.cs should be merged w/ Scripts/Projectile/Projectile.cs and renamed to Scripts/Projectile/ProjectileMovement.cs

Comment (SeijiEmery) [reply Assets/TestScript/BulletMovement.cs]: should add destruction timer:
`public bool destroySelfAfterTimeout = false`
`public float timeoutLifetime = 10f`

Track timeout using Start() and Time.time, eg. add a private variable private float `destructionTime`, init to `destructionTime = Time.time + timeoutLifetime`, and trigger destruction iff `destructionTime <= Time.time`

destroy on timeout iff `destroySelfAfterTimeout == true`.

either just add this directly, or if you find a better way to implement this do that

Comment (SeijiEmery) [reply Assets/TestScript/BulletMovement.cs]: should add a public `Destroy()` method that can be called externally and move destruction there

Comment (SeijiEmery) [reply Assets/TestScript/BulletMovement.cs]: should add a `public bool destroySelfOnFirstCollision = false` variable and use that to destroy this object on collision iff that is set to true

**Review by SeijiEmery — CHANGES_REQUESTED** [review]
Needs some changes to BulletMotion.cs; after that can be merged.

Test scenes + scripts are fine until we get the other scripts needed for ship controls + movement up (which can / should be another feature branch + PR); we'll update / refactor this when those are done.

File: Assets/TestScript/BulletMovement.cs (hunk 1 above)
Comment (SeijiEmery) [reply Assets/TestScript/BulletMovement.cs]: bunch of stuff, but if all of these are done I'll say that Scripts/Projectile/ProjectileMotion.cs is done.

File: Assets/Prefabs/TempRedShip.prefab [hunk 3]
```diff
@@ -113,3 +114,19 @@ MonoBehaviour:
     type: 3}
   laserBullet: {fileID: 1060486316723860202, guid: 4ee99721e977c9045a45dc1fe6051797,
     type: 3}
+--- !u!54 &-3515665261618971134
+Rigidbody:
+  m_ObjectHideFlags: 0
```
Comment (SeijiEmery) [inline Assets/Prefabs/TempRedShip.prefab:119]: rename this => Prefabs/Sub.prefab

**Review by SeijiEmery — CHANGES_REQUESTED** [review]
sub prefab needs to be renamed. everything else looks fine

File: Assets/TestScript/CrabTestController.cs (hunk 2 above)
Comment (SeijiEmery) [reply Assets/TestScript/CrabTestController.cs]: actually, could you edit the input axes (Edit > ProjectSettings > Input):

- rename Horizontal => MoveX and Vertical => MoveZ
- add MoveY axis + add bindings for shift (down) and tab (up). Should be setup the same as the Horizontal + Vertical axes
- add TurnHorizontal axis (Q + E)
- add TurnVertical axis (duplicate bindings to shift + tab)

then you can refactor to:

`var movement = Vector3.forward * Input.GetAxis("MoveZ") + Vector3.up * ...`
`transform.Translate(movement * speed * Time.deltaTime, Space.World)`

and ditto for rotation

Comment (SeijiEmery) [reply Assets/TestScript/CrabTestController.cs]: note: if you do this and refactor all the input stuff we should be able to use both WASD and arrow keys for movement :)

File: Assets/TestScript/BulletMovement.cs (hunk 1 above)
Comment (SeijiEmery) [reply Assets/TestScript/BulletMovement.cs]: we'll still need damage, etc., but that can be a separate PR

Comment (CCD729) [reply Assets/TestScript/BulletMovement.cs]: Done. Please check bullet prefabs and ProjectileMovement.cs for details.

File: Assets/Prefabs/TempRedShip.prefab (hunk 3 above)
Comment (CCD729) [reply Assets/Prefabs/TempRedShip.prefab:119]: Renamed.

File: Assets/TestScript/CrabTestController.cs (hunk 2 above)
Comment (CCD729) [reply Assets/TestScript/CrabTestController.cs]: The new input system applied to all prefabs. MortarShells now got similar behaviors as projectiles. Optimized some code structure.

**Comment by CCD729** [comment]
Await for a second round of reviews.

File: Assets/Script/Projectiles/ProjectileMovement.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    private float destroyTime;
+
+    void Start()
+    {
+        destroyTime = Time.time + timeoutLifetime;
+    }
+    void FixedUpdate()
+    {
+        transform.Translate(Vector3.up * Time.deltaTime * speed);
```
Comment (kodomazer) [inline Assets/Script/Projectiles/ProjectileMovement.cs]: Use `Time.fixedDeltaTime` in FixedUpdate

File: Assets/Script/Projectiles/ProjectileMovement.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+using UnityEngine;
+
+// Controls projectile movement
+// Damage + triggering is applied by DamageTrigger.cs
+// TODO: implement this
+public class ProjectileMovement : MonoBehaviour
+{
+    public float speed = 2f;
+    public bool destroySelfAfterTimeout = false;
+    public bool destroySelfOnFirstCollision = false;
```
Comment (kodomazer) [inline Assets/Script/Projectiles/ProjectileMovement.cs]: This should be `destroySelfOnCollision` as its usage doesn't make it conditionally destroy itself on the second collision instead.

File: Assets/TestScript/MortarMovement.cs [hunk 6]
```diff
@@ -8,30 +8,30 @@ public class MortarMovement : MonoBehaviour
     public float speed = 2f;
     public float firingPower = 8.0f;
     public Vector3 shootingPos;
+    public bool destroySelfAfterTimeout = false;
+    public bool destroySelfOnFirstCollision = false;
```
Comment (kodomazer) [inline Assets/TestScript/MortarMovement.cs]: Same as above, rename to `destroySelfOnCollision` because it just controls if a collision destroys this.

**Review by kodomazer — CHANGES_REQUESTED** [review]
Just some minor naming changes.
I haven't opened it in the editor at all, so there might be other things that need to be changed.

Also, I tend to use:
```
[SerializedField]
private type name;
```
As my paradigm for getting variables to show up in the editor that would otherwise be private variables instead of simply making them `public`

**Review by SeijiEmery — APPROVED** [review]
Looks good; just address kodo’s change requests?

File: Assets/Script/Projectiles/ProjectileMovement.cs (hunk 4 above)
Comment (CCD729) [reply Assets/Script/Projectiles/ProjectileMovement.cs]: Replaced.

File: Assets/Script/Projectiles/ProjectileMovement.cs (hunk 5 above)
Comment (CCD729) [reply Assets/Script/Projectiles/ProjectileMovement.cs]: Changed.

File: Assets/TestScript/MortarMovement.cs (hunk 6 above)
Comment (CCD729) [reply Assets/TestScript/MortarMovement.cs]: Changed. Also, all unnecessary public vars had been set to [SerializeField]

**Review by kodomazer — APPROVED** [review]
I don't really like stacking all the variable declarations, but we can deal with that later.

**Pull request merged**